packages/common/core-utils/src/compare.ts

```diff
@@ -25,9 +25,23 @@ export const compareArrays = <T>(
 	//       '===' and 'Object.is()' tied.
 	//       Trivial acceptance adds no measurable overhead.
 	//       30% penalty vs. baseline for exported function [node 14 x64].
-	return (
-		left === right || // Trivial acceptance: 'left' and 'right' are the same instance
-		(left.length === right.length && // Trivial rejection: 'left' and 'right' are different lengths
-			left.every((leftItem, index) => comparator(leftItem, right[index], index)))
-	);
+	if (left === right) {
+		// Trivial acceptance: 'left' and 'right' are the same instance
+		return true;
+	}
+
+	if (left.length !== right.length) {
+		// Trivial rejection: 'left' and 'right' are different lengths
+		return false;
+	}
+
+	for (let index = 0; index < left.length; index++) {
+		const leftItem = left[index] as T;
+		const rightItem = right[index] as T;
+		if (!comparator(leftItem, rightItem, index)) {
+			return false;
+		}
+	}
+
+	return true;
```

Comment:
Similar here - preferable to just assert how we've accounted for the index being in-bounds rather than changing runtime behavior.

```suggestion
	return (
		left === right || // Trivial acceptance: 'left' and 'right' are the same instance
		(left.length === right.length && // Trivial rejection: 'left' and 'right' are different lengths
			// Since we've verified left.length === right.length, we know right[index] is in bounds.
			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
			left.every((leftItem, index) => comparator(leftItem, right[index]!, index)))
	);
```